lib/mongo/grid_fs/bucket.ex

```diff
@@ -0,0 +1,229 @@
+defmodule Mongo.GridFs.Bucket do
+  @moduledoc """
+
+  The bucket has some configuration options:
+    * `:chunk_size` - The chunk size in bytes. Defaults to `255*1024`
+    * `:name` - The bucket name. Defaults to `fs`
+
+  The bucket checks whether the indexes already exist before attempting to create them. The names of the
+  created indexes are "filename_1_uploadDate_1" and "files_id_1_n_1"
+
+  """
+
+  alias BSON.ObjectId
+  alias Mongo.GridFs.Bucket
+
+  ##
+  # constants used in this module
+  #
+  @files_index_name "filename_1_uploadDate_1"
+  @chunks_index_name "files_id_1_n_1"
+  @defaults [name: "fs", chunk_size: 255 * 1024]
+
+  @type t :: %__MODULE__{
+               name: String.t,
+               chunk_size: non_neg_integer,
+               topology_pid: GenServer.server
+             }
+
+  defstruct name: "fs", chunk_size: 255 * 1024, topology_pid: nil, opts: []
+
+  @doc """
+  Creates a new Bucket with a existing connection using the default values. It just contains the
+  name of the collections (fs) and the chunk size (255KB).
+
+  The bucket checks the index for both collections as well. In case of multiple
+  upload or downloads just create only one bucket and reuse it.
+
+  """
+  @spec new(GenServer.server, Keyword.t) :: Bucket.t
+  def new(topology_pid, options \\ []) do
+
+    Keyword.merge(@defaults, options)
+    |> Enum.reduce(%Bucket{topology_pid: topology_pid, opts: options}, fn {k, v}, bucket -> Map.put(bucket, k, v) end)
+    |> check_indexes
+
+  end
+
+  @doc """
+  Returns the collection name for the files collection, default is fs.files.
+  """
+  @spec files_collection_name(Bucket.t) :: String.t
+  def files_collection_name(%Bucket{name: fs}), do: "#{fs}.files"
```

Comment:
Instead of `"#{fs}.files"`, could we use `fs <> ".files"`. It should be more performant. Same goes for all similar calls.